src/arraytypes/struct.jl

```diff
@@ -33,20 +33,20 @@ isnamedtuple(T) = false
 istuple(::Type{<:Tuple}) = true
 istuple(T) = false
 
-@propagate_inbounds function Base.getindex(s::Struct{T}, i::Integer) where {T}
+@propagate_inbounds function Base.getindex(s::Struct{T,S}, i::Integer) where {T,S}
     @boundscheck checkbounds(s, i)
     NT = Base.nonmissingtype(T)
     if isnamedtuple(NT) || istuple(NT)
         if NT !== T
-            return s.validity[i] ? NT(ntuple(j->s.data[j][i], fieldcount(NT))) : missing
+            return s.validity[i] ? NT(ntuple(j->s.data[j][i], fieldcount(S))) : missing
         else
-            return NT(ntuple(j->s.data[j][i], fieldcount(NT)))
+            return NT(ntuple(j->s.data[j][i], fieldcount(S)))
```

Comment:
Not quite sure about this part. Giving it some thought

Reply:
If `S` contains more fields than `NT` we'll see an error here with the current code. The original code could also fail when the fieldcount of `NT` exceeded what was stored in `S`. As no one should be adding custom Arrow support for `Tuple` or `NamedTuple` subtypes I think this change is sensible to keep it consistent with the code below.

Reply:
Aside: as far as I'm aware the `isnamedtuple(NT) || istuple(NT)` case here only exists as the constructors for `Tuple` and `NamedTuple` expect an iterator. We could probably handle this case using `fromarrow` with `@nospecialize` to simplify the logic here